rflx/pyrflx/typevalue.py

```diff
@@ -224,39 +240,594 @@ class OpaqueValue(TypeValue):
         super().__init__(vtype)
 
-    def assign(self, value: bytes, check: bool = True) -> None:
+    def assign(self, value: bytes, length: int = 0, check: bool = True) -> None:
         self._value = value
 
-    def assign_bitvalue(self, value: str, check: bool = True) -> None:
+    def assign_bitvalue(self, value: Bitstring, length: int = 0) -> None:
+        self._value = bytes(value)
 
-        while len(value) % 8 != 0:
-            value = "0" + value
+    @property
+    def length(self) -> int:
+        self._raise_initialized()
+        return len(self._value) * 8
 
-        bytestring = b"".join(
-            [int(value[i : i + 8], 2).to_bytes(1, "big") for i in range(0, len(value), 8)]
-        )
+    @property
+    def value(self) -> bytes:
+        self._raise_initialized()
+        return self._value
+
+    @property
+    def to_bytes(self) -> bytes:
+        self._raise_initialized()
+        return self._value
+
+    @property
+    def to_bitstring(self) -> Bitstring:
+        self._raise_initialized()
+        return Bitstring(format(int.from_bytes(self._value, "big"), f"0{self.length}b"))
+
+    @property
+    def accepted_type(self) -> type:
+        return bytes
+
+    @property
+    def literals(self) -> Mapping[Name, Expr]:
+        return {}
+
+
+class ArrayValue(TypeValue):
+
+    _value: List[TypeValue]
+    _element_list_type: Type
+    _is_message_array: bool
+    _all_messages: List[model.Message]
+
+    def __init__(self, vtype: Array, all_messages: List[model.Message]) -> None:
+        super().__init__(vtype)
+        self._element_list_type = vtype.element_type
+        self._all_messages = all_messages
+        self._is_message_array = isinstance(self._element_list_type, model.Message)
+
+    def assign(self, value: List[TypeValue], length: int = 0, check: bool = True) -> None:
+
+        # check if all elements are of the same class and messages are valid
+        for v in value:
+            if (
+                self._is_message_array
+                and isinstance(v, MessageValue)
+                and isinstance(self._element_list_type, model.Message)
+            ):
+                if not v.check_model_equality(self._element_list_type):
+                    raise ValueError("members of an array must not be of different classes")
+                if not v.valid_message:
+                    raise ValueError("cannot assign array of messages: messages must be valid")
+            else:
+                if not self._is_message_array and not v.check_type_equality(
+                    self._element_list_type
+                ):
+                    raise ValueError("members of an array must not be of different classes")
+
+        self._value = value
+
+    def assign_bitvalue(self, value: Bitstring, length: int = 0) -> None:
+
+        self._value = []
+
+        if self._is_message_array:
+
+            while len(str(value)) != 0:
+
+                assert isinstance(self._element_list_type, model.Message)
```

Comment:
I suppose `assign_bitvalue` should be only used for arrays of messages. If this method is wrongly used on an other array, an exception should be raised. In general a user should never face an AssertionError when using the library.

Reply:
`assign_bitvalue` can also be used for arrays of ScalarValues. However the assertion is not needed because the loop can only be entered if `self._is_message_array` is true and that is only true if `isinstance(self._element_type, model.Message)` evaluates to true. Therefore i removed the assertion.